src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameParser.cs

```diff
@@ -0,0 +1,455 @@
+        /// Given an input text, and set of options, parses out a fully representative syntax tree 
+        /// and list of diagnostics.  Parsing should always succeed, except in the case of the stack 
+        /// overflowing.
+        /// </summary>
+        public static StackFrameTree? TryParse(VirtualCharSequence text)
+        {
+            if (text.IsDefault)
+            {
+                return null;
+            }
+
+            try
+            {
+                return new StackFrameParser(text).TryParseTree();
+            }
+            catch (InsufficientExecutionStackException)
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Constructs a <see cref="VirtualCharSequence"/> and calls <see cref="TryParse(VirtualCharSequence)"/>
+        /// </summary>
+        public static StackFrameTree? TryParse(string text)
+            => TryParse(VirtualCharSequence.Create(0, text));
+
+        /// <summary>
+        /// Attempts to parse the full tree. Returns null on malformed data
+        /// </summary>
+        private StackFrameTree? TryParseTree()
+        {
+            var (_, methodDeclaration) = TryParseMethodDeclaration();
+            if (methodDeclaration is null)
+            {
+                return null;
+            }
+
+            var fileInformationResult = TryParseFileInformation();
+            if (!fileInformationResult.Success)
+            {
+                return null;
+            }
+
+            var remainingTrivia = _lexer.TryScanRemainingTrivia();
+
+            var eolToken = CurrentCharAsToken().With(leadingTrivia: remainingTrivia.HasValue ? ImmutableArray.Create(remainingTrivia.Value) : ImmutableArray<StackFrameTrivia>.Empty);
+
+            Debug.Assert(_lexer.Position == _lexer.Text.Length);
+            Debug.Assert(eolToken.Kind == StackFrameKind.EndOfLine);
+
+            var root = new StackFrameCompilationUnit(methodDeclaration, fileInformationResult.Value, eolToken);
+
+            return new(_lexer.Text, root);
+        }
+
+        /// <summary>
+        /// Attempts to parse the full method declaration, optionally adding leading whitespace as trivia. Includes
+        /// all of the generic indicators for types, 
+        /// 
+        /// Ex: [|MyClass.MyMethod(string s)|]
+        /// </summary>
+        private ParseResult<StackFrameMethodDeclarationNode> TryParseMethodDeclaration()
+        {
+            var (success, identifierNode) = TryParseNameNode(scanAtTrivia: true);
+            if (!success)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            if (identifierNode is not StackFrameQualifiedNameNode memberAccessExpression)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            var typeArgumentsResult = TryParseTypeArguments();
+            if (!typeArgumentsResult.Success)
```

Comment:
nit:

```suggestion
            (success, var typeArgumentsResult) = TryParseTypeArguments();
            if (!success)
```

this way you jsut reuse the same 'success' value, which prevents accidental misuse.